### PR TITLE
sprinkle typehints in spack.installer

### PR DESCRIPTION
**Pull request by haampie**

### CODE REVIEW

File: lib/spack/spack/database.py [hunk 1]
```diff
@@ -710,7 +712,7 @@ def db_for_spec_hash(self, hash_key):
             if hash_key in db._data:
                 return db
 
-    def query_by_spec_hash(self, hash_key, data=None):
+    def query_by_spec_hash(self, hash_key, data=None) -> Tuple[bool, Optional[InstallRecord]]:
```
Comment (alalazo) [inline lib/spack/spack/database.py]: Is it easy to add `hash_key` and `data` in the docstring and type hints?

File: lib/spack/spack/installer.py [hunk 2]
```diff
@@ -95,7 +98,80 @@ class InstallAction:
     OVERWRITE = 2
 
 
-def _check_last_phase(pkg):
+class InstallStatus:
```
Comment (alalazo) [inline lib/spack/spack/installer.py:102]: Can we add type hints to this class too, while we are at it?

File: lib/spack/spack/installer.py [hunk 3]
```diff
@@ -95,7 +98,80 @@ class InstallAction:
...
+
+        status = "{0} {1}".format(text, self.get_progress())
+        sys.stdout.write("\033]0;Spack: {0}\007".format(status))
+        sys.stdout.flush()
+
+    def get_progress(self):
+        return "[{0}/{1}]".format(self.pkg_num, self.pkg_count)
+
+
+class TermStatusLine:
```
Comment (alalazo) [inline lib/spack/spack/installer.py:131]: Same request here

File: lib/spack/spack/installer.py [hunk 4]
```diff
@@ -104,35 +180,33 @@ def _check_last_phase(pkg):
...
-        pkg (spack.package_base.PackageBase): the package being installed
+        pkg: the package being installed
 
     Raises:
         ``BadInstallPhase`` if stop_before or last phase is invalid
     """
-    phases = pkg.builder.phases
-    if pkg.stop_before_phase and pkg.stop_before_phase not in phases:
-        raise BadInstallPhase(pkg.name, pkg.stop_before_phase)
+    phases = pkg.builder.phases  # type: ignore[attr-defined]
```
Comment (alalazo) [inline lib/spack/spack/installer.py:185]: :cry: I wonder how hard it is to let mypy check through the metaclasses we have. Definitely for another PR.

File: lib/spack/spack/installer.py [hunk 5]
```diff
@@ -194,19 +268,19 @@ def _do_fake_install(pkg):
     dump_packages(pkg.spec, packages_dir)
 
 
-def _packages_needed_to_bootstrap_compiler(compiler, architecture, pkgs):
+def _packages_needed_to_bootstrap_compiler(
+    compiler: "spack.spec.CompilerSpec", architecture: "spack.spec.ArchSpec", pkgs: list
+):
```
Comment (alalazo) [inline lib/spack/spack/installer.py]: Missing return type

File: lib/spack/spack/installer.py [hunk 6]
```diff
@@ -293,27 +367,27 @@ def print_install_test_log(pkg: "spack.package_base.PackageBase"):
     pkg.tester.print_log_path()
 
 
-def _print_timer(pre, pkg_id, timer):
+def _print_timer(pre, pkg_id, timer: timer.Timer):
```
Comment (alalazo) [inline lib/spack/spack/installer.py]: Here we miss some arguments and `-> None`

File: lib/spack/spack/installer.py [hunk 7]
```diff
@@ -377,19 +451,23 @@ def _process_external_package(pkg, explicit):
 
 
 def _process_binary_cache_tarball(
-    pkg, explicit, unsigned, mirrors_for_spec=None, timer=timer.NULL_TIMER
+    pkg: "spack.package_base.PackageBase",
+    explicit: bool,
+    unsigned: bool,
+    mirrors_for_spec: Optional[list] = None,
+    timer: timer.BaseTimer = timer.NULL_TIMER,
```
Comment (alalazo) [inline lib/spack/spack/installer.py:454]: Missing return type

File: lib/spack/spack/installer.py [hunk 8]
```diff
@@ -465,7 +548,7 @@ def combine_phase_logs(phase_log_files, log_path):
                 shutil.copyfileobj(phase_log, log_file)
 
 
-def dump_packages(spec, path):
+def dump_packages(spec: "spack.spec.Spec", path):
```
Comment (alalazo) [inline lib/spack/spack/installer.py]:
```suggestion
def dump_packages(spec: "spack.spec.Spec", path: str) -> None:
```

File: lib/spack/spack/installer.py [hunk 9]
```diff
@@ -523,20 +606,20 @@ def dump_packages(spec, path):
             fs.install_tree(source_pkg_dir, dest_pkg_dir)
 
 
-def get_dependent_ids(spec):
+def get_dependent_ids(spec: "spack.spec.Spec"):
```
Comment (alalazo) [inline lib/spack/spack/installer.py]: Missing return type

**Review by alalazo — CHANGES_REQUESTED** [review]
I think we miss the return types in many places. Can you double check that?

File: lib/spack/spack/database.py (hunk 1 above)
Comment (haampie) [reply lib/spack/spack/database.py]: also added to _data in the DB... weird that it wasn't yet :p

**Review by alalazo — APPROVED** [review]

**Pull request merged**